src/audio/rtnr/rtnr.c

```diff
@@ -0,0 +1,546 @@
+#define SpkNum				2
+
+#if SAMPLING_RATE == 48000
+#define BLOCKLENGTH	768
+#elif SAMPLING_RATE == 16000
+#define BLOCKLENGTH	256
+#elif SAMPLING_RATE == 8000
+#define BLOCKLENGTH	128
+#endif // SAMPLING_RATE
+
+
+#define RTNR_BLK_LENGTH 4
+#define RTNR_MAX_SOURCES 1 /* Microphone stream */
+
+/*	MIC input is at index 0 */
+#define RTNR_MIC_SOURCE_INDEX (0)
+
+static const struct comp_driver comp_rtnr;
+
+/** \brief RTNR processing functions map item. */
+struct rtnr_func_map {
+	enum sof_ipc_frame fmt; /**< source frame format */
+	rtnr_func func; /**< processing function */
+};
+
+#if CONFIG_FORMAT_S16LE
+static void rtnr_s16_default(struct comp_dev *dev, const struct audio_stream **sources,
+			    struct audio_stream *sink, int frames);
+#endif /* CONFIG_FORMAT_S16LE */
+
+#if CONFIG_FORMAT_S24LE
+static void rtnr_s24_default(struct comp_dev *dev, const struct audio_stream **sources,
+			    struct audio_stream *sink, int frames);
+#endif /* CONFIG_FORMAT_S24LE */
+
+#if CONFIG_FORMAT_S32LE
+static void rtnr_s32_default(struct comp_dev *dev, const struct audio_stream **sources,
+			    struct audio_stream *sink, int frames);
+#endif /* CONFIG_FORMAT_S32LE */
+
+/* Processing functions table */
+const struct rtnr_func_map rtnr_fnmap[] = {
+#if CONFIG_FORMAT_S16LE
+	{ SOF_IPC_FRAME_S16_LE, rtnr_s16_default },
+#endif // CONFIG_FORMAT_S16LE
+#if CONFIG_FORMAT_S24LE
+	{ SOF_IPC_FRAME_S24_4LE, rtnr_s24_default },
+#endif // CONFIG_FORMAT_S24LE
+#if CONFIG_FORMAT_S32LE
+	{ SOF_IPC_FRAME_S32_LE, rtnr_s32_default },
+#endif// CONFIG_FORMAT_S32LE
+};
+
+const size_t rtnr_fncount = ARRAY_SIZE(rtnr_fnmap);
+
+/* UUID 5c7ca334-e15d-11eb-ba80-0242ac130004 */
+DECLARE_SOF_RT_UUID("rtnr", rtnr_uuid, 0x5c7ca334, 0xe15d, 0x11eb, 0xba, 0x80,
+		    0x02, 0x42, 0xac, 0x13, 0x00, 0x04);
+
+DECLARE_TR_CTX(rtnr_tr, SOF_UUID(rtnr_uuid), LOG_LEVEL_INFO);
+
+/* Generic processing */
+
+
+void myPrintf(int a, int b, int c, int d, int e)
+{
+	if (a == 0xa)
+		comp_cl_info(&comp_rtnr, "myPrintf 1st=%08x, 2nd=%08x, 3rd=%08x, 4st=%08x",
+			b, c, d, e);
+	else if	(a == 0xb)
+		comp_cl_info(&comp_rtnr, "myPrintf 1st=%08x, 2nd=%08x, 3rd=%08x, 4st=%08x",
+			b, c, d, e);
+	else if	(a == 0xc)
+		comp_cl_warn(&comp_rtnr, "myPrintf 1st=%08x, 2nd=%08x, 3rd=%08x, 4st=%08x",
+			b, c, d, e);
+	else if	(a == 0xd)
+		comp_cl_dbg(&comp_rtnr, "myPrintf 1st=%08x, 2nd=%08x, 3rd=%08x, 4st=%08x",
+			b, c, d, e);
+	else if	(a == 0xe)
+		comp_cl_err(&comp_rtnr, "myPrintf 1st=%08x, 2nd=%08x, 3rd=%08x, 4st=%08x",
+			b, c, d, e);
+}
+
+void *rtk_rballoc(unsigned int flags, unsigned int caps, unsigned int bytes)
+{
+	return rballoc(flags, caps, bytes);
+}
+
+void rtk_rfree(void *ptr)
+{
+	rfree(ptr);
+}
+
+#if CONFIG_FORMAT_S16LE
+
+static void rtnr_s16_default(struct comp_dev *dev, const struct audio_stream **sources,
+			    struct audio_stream *sink, int frames)
+
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	RTKMA_API_S16_Default((void *)&cd->rtk_agl, sources, sink, frames,
```

Comment:
Is RTKMA_API_S16_Default a function in the proprietary library? If yes I think the interface should be simplified, not dependent on any SOF struct, e.g. audio_stream.

Reply:
Yes it is a function in the proprietary library. The internal usage of audio_stream might be changed in the future, but we want to keep the interface consistent. That's why we pass the pointer to audio_stream to the library.

Reply:
What do you mean "we want to keep the interface consistent"?

Reply:
The processing function may use some specific information in audio_stream for different applications, and we want the interface to the wrapper code to be identical.